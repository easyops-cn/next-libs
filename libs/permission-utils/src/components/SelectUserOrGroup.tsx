import React from "react";
import { Select } from "antd";
import { filter, isEmpty } from "lodash";
import {
  SelectUserOrGroupProps,
  SelectUserOrGroupState,
  User,
} from "../interfaces";
import { InstanceApi_postSearch } from "@next-sdk/cmdb-sdk";
import { handleHttpError } from "@next-core/brick-kit";
import { LabeledValue } from "antd/lib/select";

export class SelectUserOrGroup extends React.Component<
  SelectUserOrGroupProps,
  SelectUserOrGroupState
> {
  constructor(props: SelectUserOrGroupProps) {
    super(props);
    this.state = {
      users: [],
      userGroups: [],
    };
  }
  fetchUser() {
    return InstanceApi_postSearch("USER", {
      query: {},
      fields: {
        instanceId: true,
        name: true,
      },
    });
  }
  fetchUserGroup() {
    return InstanceApi_postSearch("USER_GROUP", {
      query: {},
      fields: {
        instanceId: true,
        name: true,
      },
    });
  }
  async componentDidMount() {
    try {
      const [users, groups] = await Promise.all([
        this.fetchUser(),
        this.fetchUserGroup(),
      ]);
      this.setState({
        users: users.list as User[],
        userGroups: groups.list as User[],
      });
    } catch (e) {
      handleHttpError(e);
    }
  }
  handleUsersChange = (value: LabeledValue[]) => {
    this.props.handleUsersChange(value);
  };
  filterOpts = (currentUsers: string[], allUsers: User[]) => {
    return filter(allUsers, (item) => currentUsers.includes(item.name));
  };
  render(): React.ReactNode {
    const { currentUsers } = this.props;
    const { users: allUsers, userGroups: allUserGroups } = this.state;
    const users = isEmpty(currentUsers)
      ? allUsers
      : this.filterOpts(currentUsers, allUsers);
    const userGroups = isEmpty(currentUsers)
      ? allUserGroups
      : this.filterOpts(currentUsers, allUserGroups);
    return (
      <Select
        labelInValue
        style={{ width: "100%" }}
        mode="multiple"
        placeholder="选择用户（组）"
        onChange={this.handleUsersChange}
        optionFilterProp="label"
      >
        <Select.OptGroup label="用户">
          {users.map((item: any) => (
            <Select.Option
              value={item.instanceId}
              key={item.instanceId}
              label={item.name}
            >
              {item.name}
            </Select.Option>
          ))}
        </Select.OptGroup>
        <Select.OptGroup label="用户组">
          {userGroups.map((item: any) => (
            <Select.Option
              value={item.instanceId}
              key={item.instanceId}
              label={item.name}
            >
              {item.name}
            </Select.Option>
          ))}
        </Select.OptGroup>
      </Select>
    );
  }
}
