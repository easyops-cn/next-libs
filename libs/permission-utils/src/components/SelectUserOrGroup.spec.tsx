import React from "react";
import { shallow } from "enzyme";
import { SelectUserOrGroup } from "./SelectUserOrGroup";
import { Select } from "antd";
import { exportDefaultDeclaration } from "@babel/types";
describe("SelectUserOrGroup", () => {
  const handleUsersChange = jest.fn();
  const props = { handleUsersChange };
  const wrapper = shallow(<SelectUserOrGroup {...props} />);
  const instance = wrapper.instance() as SelectUserOrGroup;
  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Select)).toHaveLength(1);
  });

  it("should handle change", () => {
    const spy = jest.spyOn(props, "handleUsersChange");
    instance.handleUsersChange([
      { label: "user", value: "12345", key: "12345" },
    ]);
    expect(spy).toHaveBeenCalled();
  });
  it("should work when call filterOpts function", () => {
    const allUsers = [
      {
        instanceId: "abc",
        name: "lightjiao",
      },
      {
        instanceId: "easyops",
        name: "easyops",
      },
      {
        instanceId: "willniu",
        name: "1234567",
      },
      {
        instanceId: "test",
        name: "12345678",
      },
    ];
    expect(instance.filterOpts(["lightjiao", "easyops"], allUsers)).toEqual([
      {
        instanceId: "abc",
        name: "lightjiao",
      },
      {
        instanceId: "easyops",
        name: "easyops",
      },
    ]);
  });
});
