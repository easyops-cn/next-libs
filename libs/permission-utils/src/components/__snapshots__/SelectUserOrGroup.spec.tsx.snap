// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectUserOrGroup should render 1`] = `
<Select
  choiceTransitionName="zoom"
  labelInValue={true}
  mode="multiple"
  onChange={[Function]}
  placeholder="选择用户（组）"
  showSearch={false}
  style={
    Object {
      "width": "100%",
    }
  }
  transitionName="slide-up"
>
  <OptGroup
    label="用户"
  />
  <OptGroup
    label="用户组"
  />
</Select>
`;
